File: apps/web-teacher/lib/marketplace-api.ts
/**
 * Marketplace API Client for Teacher Portal
 *
 * API client for accessing marketplace items that are installed
 * and approved for the teacher's school/district.
 */

import { getServiceUrl } from './env-utils';

const API_BASE = getServiceUrl(
  'NEXT_PUBLIC_MARKETPLACE_API_URL',
  'http://localhost:3007/api',
  'Marketplace API'
);

// ============================================================================
// Types
// ============================================================================

export type MarketplaceItemType = 'CONTENT_PACK' | 'EMBEDDED_TOOL';

export interface MarketplaceLibraryItem {
  id: string;
  title: string;
  slug: string;
  description: string;
  iconUrl?: string;
  itemType: MarketplaceItemType;
  vendor: {
    id: string;
    name: string;
    logoUrl?: string;
  };
  version: {
    id: string;
    version: string;
  };
  installation: {
    id: string;
    status: string;
    gradeBandConfig?: string[];
    config?: Record<string, unknown>;
  };
  subjects: string[];
  gradeBands: string[];
  rating: number;
  reviewCount: number;
  safetyCertified: boolean;
}

export interface ClassroomContent {
  id: string;
  classroomId: string;
  marketplaceItemId: string;
  addedAt: string;
  addedBy: string;
  item: {
    id: string;
    title: string;
    slug: string;
    iconUrl?: string;
    itemType: MarketplaceItemType;
    vendor: {
      name: string;
    };
  };
}

// ============================================================================
// API Functions
// ============================================================================

/**
 * Get all marketplace items available to the teacher (installed and approved by admin)
 */
export async function getTeacherLibrary(
  teacherId: string,
  filters?: {
    type?: MarketplaceItemType;
    subject?: string;
    gradeBand?: string;
    search?: string;
  }
): Promise<{ data: MarketplaceLibraryItem[] }> {
  const params = new URLSearchParams();
  if (filters?.type) params.set('type', filters.type);
  if (filters?.subject) params.set('subject', filters.subject);
  if (filters?.gradeBand) params.set('gradeBand', filters.gradeBand);
  if (filters?.search) params.set('q', filters.search);

  const res = await fetch(`${API_BASE}/teacher/${teacherId}/library?${params.toString()}`, {
    headers: { 'Content-Type': 'application/json' },
    cache: 'no-store',
  });

  if (!res.ok) {
    throw new Error(`Failed to fetch teacher library: ${res.statusText}`);
  }

  return res.json();
}

/**
 * Get marketplace item details
 */
export async function getLibraryItemBySlug(slug: string): Promise<MarketplaceLibraryItem | null> {
  const res = await fetch(`${API_BASE}/catalog/${slug}`, {
    headers: { 'Content-Type': 'application/json' },
    cache: 'no-store',
  });

  if (res.status === 404) return null;
  if (!res.ok) {
    throw new Error(`Failed to fetch item: ${res.statusText}`);
  }

  return res.json();
}

/**
 * Get content items added to a specific classroom
 */
export async function getClassroomContent(
  classroomId: string
): Promise<{ data: ClassroomContent[] }> {
  const res = await fetch(`${API_BASE}/classrooms/${classroomId}/content`, {
    headers: { 'Content-Type': 'application/json' },
    cache: 'no-store',
  });

  if (!res.ok) {
    throw new Error(`Failed to fetch classroom content: ${res.statusText}`);
  }

  return res.json();
}

/**
 * Add a marketplace item to a classroom
 */
export async function addContentToClassroom(
  classroomId: string,
  marketplaceItemId: string,
  teacherId: string
): Promise<ClassroomContent> {
  const res = await fetch(`${API_BASE}/classrooms/${classroomId}/content`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ marketplaceItemId, addedBy: teacherId }),
  });

  if (!res.ok) {
    throw new Error(`Failed to add content to classroom: ${res.statusText}`);
  }

  return res.json();
}

/**
 * Remove a marketplace item from a classroom
 */
export async function removeContentFromClassroom(
  classroomId: string,
  contentId: string
): Promise<void> {
  const res = await fetch(`${API_BASE}/classrooms/${classroomId}/content/${contentId}`, {
    method: 'DELETE',
    headers: { 'Content-Type': 'application/json' },
  });

  if (!res.ok) {
    throw new Error(`Failed to remove content from classroom: ${res.statusText}`);
  }
}

/**
 * Get teacher's classrooms for adding content
 * Enterprise UI Audit: RE-AUDIT-AUTH-001 - Added to replace mock classroom data
 */
export interface TeacherClassroom {
  id: string;
  name: string;
  gradeBand: string;
  studentCount: number;
}

export async function getTeacherClassrooms(
  teacherId: string
): Promise<{ data: TeacherClassroom[] }> {
  const res = await fetch(`${API_BASE}/teacher/${teacherId}/classrooms`, {
    headers: { 'Content-Type': 'application/json' },
    cache: 'no-store',
  });

  if (!res.ok) {
    throw new Error(`Failed to fetch teacher classrooms: ${res.statusText}`);
  }

  return res.json();
}

// ============================================================================
// Helper Functions
// ============================================================================

export function getItemTypeLabel(type: MarketplaceItemType): string {
  const labels: Record<MarketplaceItemType, string> = {
    CONTENT_PACK: 'Content Pack',
    EMBEDDED_TOOL: 'Embedded Tool',
  };
  return labels[type] || type;
}

export function getItemTypeIcon(type: MarketplaceItemType): string {
  return type === 'CONTENT_PACK' ? '📚' : '🔧';
}

export function getSubjectLabel(subject: string): string {
  const labels: Record<string, string> = {
    MATHEMATICS: 'Mathematics',
    ENGLISH_LANGUAGE_ARTS: 'English Language Arts',
    SCIENCE: 'Science',
    SOCIAL_STUDIES: 'Social Studies',
    FOREIGN_LANGUAGE: 'Foreign Language',
    ART: 'Art',
    MUSIC: 'Music',
    PHYSICAL_EDUCATION: 'Physical Education',
    COMPUTER_SCIENCE: 'Computer Science',
    HEALTH: 'Health',
  };
  return labels[subject] || subject;
}

export function getGradeBandLabel(gradeBand: string): string {
  const labels: Record<string, string> = {
    PRE_K: 'Pre-K',
    K_2: 'K–2',
    GRADES_3_5: '3–5',
    GRADES_6_8: '6–8',
    GRADES_9_12: '9–12',
  };
  return labels[gradeBand] || gradeBand;
}

// ============================================================================
// Partner Content API Functions
// ============================================================================

export interface PartnerContentItem {
  id: string;
  slug: string;
  title: string;
  shortDescription: string;
  itemType: MarketplaceItemType;
  subjects: string[];
  gradeBands: string[];
  iconUrl?: string;
  vendor: {
    id: string;
    slug: string;
    name: string;
    logoUrl?: string;
  };
  license: {
    id: string;
    status: string;
    seatLimit: number | null;
    seatsUsed: number;
    validUntil: string | null;
  };
  loCount: number;
  accessibilityTags: string[];
  safetyTags: string[];
}

export interface EntitledContentResponse {
  data: PartnerContentItem[];
  pagination: {
    total: number;
    limit: number;
    offset: number;
    hasMore: boolean;
  };
}

/**
 * Get partner content that the tenant is entitled to
 * Used by the partner content tab in the content picker
 */
export async function getEntitledPartnerContent(
  tenantId: string,
  options?: {
    schoolId?: string;
    classroomId?: string;
    gradeBand?: string;
    subject?: string;
    itemType?: MarketplaceItemType;
    limit?: number;
    offset?: number;
  }
): Promise<EntitledContentResponse> {
  const res = await fetch(`${API_BASE}/internal/entitlements/entitled-content`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      tenantId,
      schoolId: options?.schoolId,
      classroomId: options?.classroomId,
      gradeBand: options?.gradeBand,
      subject: options?.subject,
      itemType: options?.itemType,
      limit: options?.limit ?? 50,
      offset: options?.offset ?? 0,
    }),
    cache: 'no-store',
  });

  if (!res.ok) {
    throw new Error(`Failed to fetch entitled partner content: ${res.statusText}`);
  }

  return res.json();
}

/**
 * Check if tenant has entitlement to specific LOs
 * Used for filtering content before display
 */
export async function checkEntitlements(
  tenantId: string,
  loIds: string[],
  options?: {
    learnerId?: string;
    schoolId?: string;
    classroomId?: string;
    gradeBand?: string;
  }
): Promise<{
  entitled: string[];
  denied: Array<{ loId: string; reason: string }>;
  summary: {
    totalRequested: number;
    totalEntitled: number;
    totalDenied: number;
  };
}> {
  const res = await fetch(`${API_BASE}/internal/entitlements/batch-check`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      tenantId,
      loIds,
      learnerId: options?.learnerId,
      schoolId: options?.schoolId,
      classroomId: options?.classroomId,
      gradeBand: options?.gradeBand,
    }),
    cache: 'no-store',
  });

  if (!res.ok) {
    throw new Error(`Failed to check entitlements: ${res.statusText}`);
  }

  return res.json();
}

/**
 * Get entitled LO IDs for partner content
 */
export async function getEntitledLoIds(
  tenantId: string,
  options?: {
    schoolId?: string;
    gradeBand?: string;
  }
): Promise<{ loIds: string[]; count: number }> {
  const res = await fetch(`${API_BASE}/internal/entitlements/entitled-los`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      tenantId,
      schoolId: options?.schoolId,
      gradeBand: options?.gradeBand,
    }),
    cache: 'no-store',
  });

  if (!res.ok) {
    throw new Error(`Failed to fetch entitled LO IDs: ${res.statusText}`);
  }

  return res.json();
}

File: libs/flutter-common/lib/api/marketplace_api.dart
/// Marketplace API Client
///
/// API client for marketplace entitlements and partner content.
library;

import 'package:dio/dio.dart';

import 'api_client.dart';

/// Types
enum MarketplaceItemType {
  contentPack,
  embeddedTool,
}

/// Partner content item from marketplace
class PartnerContentItem {
  final String id;
  final String slug;
  final String title;
  final String shortDescription;
  final MarketplaceItemType itemType;
  final List<String> subjects;
  final List<String> gradeBands;
  final String? iconUrl;
  final PartnerVendor vendor;
  final PartnerLicense license;
  final int loCount;
  final List<String> accessibilityTags;
  final List<String> safetyTags;

  const PartnerContentItem({
    required this.id,
    required this.slug,
    required this.title,
    required this.shortDescription,
    required this.itemType,
    required this.subjects,
    required this.gradeBands,
    this.iconUrl,
    required this.vendor,
    required this.license,
    required this.loCount,
    required this.accessibilityTags,
    required this.safetyTags,
  });

  factory PartnerContentItem.fromJson(Map<String, dynamic> json) {
    return PartnerContentItem(
      id: json['id'] as String,
      slug: json['slug'] as String,
      title: json['title'] as String,
      shortDescription: json['shortDescription'] as String,
      itemType: json['itemType'] == 'CONTENT_PACK'
          ? MarketplaceItemType.contentPack
          : MarketplaceItemType.embeddedTool,
      subjects: List<String>.from(json['subjects'] ?? []),
      gradeBands: List<String>.from(json['gradeBands'] ?? []),
      iconUrl: json['iconUrl'] as String?,
      vendor: PartnerVendor.fromJson(json['vendor'] as Map<String, dynamic>),
      license: PartnerLicense.fromJson(json['license'] as Map<String, dynamic>),
      loCount: json['loCount'] as int? ?? 0,
      accessibilityTags: List<String>.from(json['accessibilityTags'] ?? []),
      safetyTags: List<String>.from(json['safetyTags'] ?? []),
    );
  }
}

class PartnerVendor {
  final String id;
  final String slug;
  final String name;
  final String? logoUrl;

  const PartnerVendor({
    required this.id,
    required this.slug,
    required this.name,
    this.logoUrl,
  });

  factory PartnerVendor.fromJson(Map<String, dynamic> json) {
    return PartnerVendor(
      id: json['id'] as String,
      slug: json['slug'] as String,
      name: json['name'] as String,
      logoUrl: json['logoUrl'] as String?,
    );
  }
}

class PartnerLicense {
  final String id;
  final String status;
  final int? seatLimit;
  final int seatsUsed;
  final DateTime? validUntil;

  const PartnerLicense({
    required this.id,
    required this.status,
    this.seatLimit,
    required this.seatsUsed,
    this.validUntil,
  });

  factory PartnerLicense.fromJson(Map<String, dynamic> json) {
    return PartnerLicense(
      id: json['id'] as String,
      status: json['status'] as String,
      seatLimit: json['seatLimit'] as int?,
      seatsUsed: json['seatsUsed'] as int? ?? 0,
      validUntil: json['validUntil'] != null
          ? DateTime.parse(json['validUntil'] as String)
          : null,
    );
  }

  bool get hasSeatsAvailable =>
      seatLimit == null || seatsUsed < seatLimit!;
}

class EntitledContentResponse {
  final List<PartnerContentItem> data;
  final int total;
  final bool hasMore;

  const EntitledContentResponse({
    required this.data,
    required this.total,
    required this.hasMore,
  });

  factory EntitledContentResponse.fromJson(Map<String, dynamic> json) {
    final pagination = json['pagination'] as Map<String, dynamic>;
    return EntitledContentResponse(
      data: (json['data'] as List)
          .map((e) => PartnerContentItem.fromJson(e as Map<String, dynamic>))
          .toList(),
      total: pagination['total'] as int,
      hasMore: pagination['hasMore'] as bool,
    );
  }
}

class EntitlementCheckResult {
  final List<String> entitled;
  final List<DeniedEntitlement> denied;
  final int totalRequested;
  final int totalEntitled;

  const EntitlementCheckResult({
    required this.entitled,
    required this.denied,
    required this.totalRequested,
    required this.totalEntitled,
  });

  factory EntitlementCheckResult.fromJson(Map<String, dynamic> json) {
    final summary = json['summary'] as Map<String, dynamic>;
    return EntitlementCheckResult(
      entitled: List<String>.from(json['entitled'] ?? []),
      denied: (json['denied'] as List?)
              ?.map((e) => DeniedEntitlement.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      totalRequested: summary['totalRequested'] as int,
      totalEntitled: summary['totalEntitled'] as int,
    );
  }
}

class DeniedEntitlement {
  final String loId;
  final String reason;

  const DeniedEntitlement({
    required this.loId,
    required this.reason,
  });

  factory DeniedEntitlement.fromJson(Map<String, dynamic> json) {
    return DeniedEntitlement(
      loId: json['loId'] as String,
      reason: json['reason'] as String,
    );
  }
}

/// Marketplace API client for entitlement checks and partner content.
class MarketplaceApiClient {
  MarketplaceApiClient({Dio? dio}) : _dio = dio ?? AivoApiClient.instance.dio;

  final Dio _dio;

  static const String _internalPath = '/internal/entitlements';

  /// Get entitled partner content for the current tenant.
  ///
  /// Returns content packs and tools that the tenant has active licenses for.
  Future<EntitledContentResponse> getEntitledPartnerContent({
    required String tenantId,
    String? schoolId,
    String? classroomId,
    String? gradeBand,
    String? subject,
    MarketplaceItemType? itemType,
    int limit = 50,
    int offset = 0,
  }) async {
    final response = await _dio.post<Map<String, dynamic>>(
      '$_internalPath/entitled-content',
      data: {
        'tenantId': tenantId,
        if (schoolId != null) 'schoolId': schoolId,
        if (classroomId != null) 'classroomId': classroomId,
        if (gradeBand != null) 'gradeBand': gradeBand,
        if (subject != null) 'subject': subject,
        if (itemType != null)
          'itemType': itemType == MarketplaceItemType.contentPack
              ? 'CONTENT_PACK'
              : 'EMBEDDED_TOOL',
        'limit': limit,
        'offset': offset,
      },
    );

    return EntitledContentResponse.fromJson(response.data!);
  }

  /// Batch check entitlements for multiple LOs.
  ///
  /// Returns which LOs the tenant is entitled to access.
  Future<EntitlementCheckResult> batchCheckEntitlements({
    required String tenantId,
    required List<String> loIds,
    String? learnerId,
    String? schoolId,
    String? classroomId,
    String? gradeBand,
  }) async {
    final response = await _dio.post<Map<String, dynamic>>(
      '$_internalPath/batch-check',
      data: {
        'tenantId': tenantId,
        'loIds': loIds,
        if (learnerId != null) 'learnerId': learnerId,
        if (schoolId != null) 'schoolId': schoolId,
        if (classroomId != null) 'classroomId': classroomId,
        if (gradeBand != null) 'gradeBand': gradeBand,
      },
    );

    return EntitlementCheckResult.fromJson(response.data!);
  }

  /// Get all entitled LO IDs from partner content.
  Future<List<String>> getEntitledLoIds({
    required String tenantId,
    String? schoolId,
    String? gradeBand,
  }) async {
    final response = await _dio.post<Map<String, dynamic>>(
      '$_internalPath/entitled-los',
      data: {
        'tenantId': tenantId,
        if (schoolId != null) 'schoolId': schoolId,
        if (gradeBand != null) 'gradeBand': gradeBand,
      },
    );

    return List<String>.from(response.data!['loIds'] ?? []);
  }

  /// Check if a single LO is entitled for the tenant.
  Future<bool> isLoEntitled({
    required String tenantId,
    required String loId,
    String? learnerId,
    String? schoolId,
    String? classroomId,
    String? gradeBand,
  }) async {
    final response = await _dio.post<Map<String, dynamic>>(
      '$_internalPath/check',
      data: {
        'tenantId': tenantId,
        'loId': loId,
        if (learnerId != null) 'learnerId': learnerId,
        if (schoolId != null) 'schoolId': schoolId,
        if (classroomId != null) 'classroomId': classroomId,
        if (gradeBand != null) 'gradeBand': gradeBand,
      },
    );

    return response.data!['isAllowed'] as bool;
  }
}
